staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -513,3 +530,91 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
 func (u unionTransformers) TransformToStorage(ctx context.Context, data []byte, dataCtx value.Context) (out []byte, err error) {
 	return u[0].TransformToStorage(ctx, data, dataCtx)
 }
+
+// DynamicTransformers holds transformers that are dynamically updated.
+type DynamicTransformers struct {
+	transformTracker *atomic.Value
+}
+
+type transformTracker struct {
+	transformerOverrides map[schema.GroupResource]value.Transformer
+	closeTransformers    context.CancelFunc
+}
+
+type notificationChannel struct {
+	once sync.Once
+	ch   chan struct{}
+}
+
+// NewDynamicTransformers initializes DynamicTransformers when api-server starts.
+func NewDynamicTransformers(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+) *DynamicTransformers {
+	dynamicTransformers := &DynamicTransformers{
+		transformTracker: &atomic.Value{},
+	}
+
+	tracker := &transformTracker{
+		transformerOverrides: transformerOverrides,
+		closeTransformers:    closeTransformers,
+	}
+	dynamicTransformers.transformTracker.Store(tracker)
+
+	return dynamicTransformers
+}
+
+// TransformerForResource returns the transformer for the given resource.
+func (d *DynamicTransformers) TransformerForResource(resource schema.GroupResource) value.Transformer {
+	return &resourceTransformer{
+		resource:         resource,
+		transformTracker: d.transformTracker,
+	}
+}
+
+// Set sets the transformer overrides. This method is not go routine safe and must only be called by the same, single caller throughout the lifetime of this object.
+func (d *DynamicTransformers) Set(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+	maxTimeout time.Duration,
+) {
+	// store new values
+	newTransformTracker := &transformTracker{
+		transformerOverrides: transformerOverrides,
+		closeTransformers:    closeTransformers,
+	}
+
+	// update new transformer overrides
+	oldTransformTracker := d.transformTracker.Swap(newTransformTracker).(*transformTracker)
+
+	// close old transformers once we wait twice as long as the max grpc request timeout
+	// this ensures that we wait for any in-flight requests to finish
+	// by the time we spawn this go routine, the new transformers have already been set and will be used for new requests
+	go func() {
+		time.Sleep(maxTimeout * 2)
```

Comment:
This should come from the old config:

```suggestion
		time.Sleep(oldTransformTracker.maxTimeout * 2)
```